bonsai/ethereum-relay/src/downloader/proxy_callback_proof_request_stream.rs

```diff
@@ -11,61 +11,127 @@
 
-use std::sync::Arc;
-
-use anyhow::anyhow;
 use bonsai_proxy_contract::CallbackRequestFilter;
 use ethers::{
-    providers::{Middleware, PubsubClient},
-    types::Address,
+    core::k256::ecdsa::SigningKey,
+    middleware::SignerMiddleware,
+    prelude::{signer::SignerMiddlewareError, *},
+    providers::{Middleware, Provider, PubsubClient, SubscriptionStream, Ws},
+    types::{Address, Log},
 };
 use futures::StreamExt;
+use tracing::{debug, error, warn};
 
-use crate::{api::error::Error, downloader::event_processor::EventProcessor};
+use crate::{api::error::Error, downloader::event_processor::EventProcessor, EthersClientConfig};
 
 pub(crate) struct ProxyCallbackProofRequestStream<
-    M: Middleware,
     EP: EventProcessor<Event = CallbackRequestFilter> + Sync + Send,
 > {
-    ethers_client: Arc<M>,
+    client_config: EthersClientConfig,
     proxy_contract_address: Address,
     event_processor: EP,
 }
 
-impl<M: Middleware, EP: EventProcessor<Event = CallbackRequestFilter> + Sync + Send>
-    ProxyCallbackProofRequestStream<M, EP>
-where
-    <M as Middleware>::Provider: PubsubClient,
+impl<EP: EventProcessor<Event = CallbackRequestFilter> + Sync + Send>
+    ProxyCallbackProofRequestStream<EP>
 {
     pub(crate) fn new(
-        ethers_client: Arc<M>,
+        client_config: EthersClientConfig,
         proxy_contract_address: Address,
         event_processor: EP,
-    ) -> ProxyCallbackProofRequestStream<M, EP> {
+    ) -> ProxyCallbackProofRequestStream<EP> {
         Self {
-            ethers_client,
+            client_config,
             proxy_contract_address,
             event_processor,
         }
     }
 
-    pub(crate) async fn run(self) -> Result<(), crate::api::error::Error> {
-        let event_name = "CallbackRequest(address,bytes32,bytes,address,bytes4,uint64)";
+    pub(crate) async fn run(self) -> Result<(), Error> {
+        const WAIT_DURATION: Duration = Duration::from_secs(5);
+        const EVENT_NAME: &str = "CallbackRequest(address,bytes32,bytes,address,bytes4,uint64)";
+        const MAX_RETRIES: usize = 60;
         let filter = ethers::types::Filter::new()
             .address(self.proxy_contract_address)
-            .event(event_name);
-        let mut proxy_stream = self
-            .ethers_client
-            .subscribe_logs(&filter)
-            .await
-            .map_err(|err| Error::Unspecified(anyhow!("{}", err)))?;
+            .event(EVENT_NAME);
+        let mut client = self.client_config.get_client().await?;
+        let mut recreate_client = false;
 
-        while let Some(log) = proxy_stream.next().await {
-            let event: CallbackRequestFilter = ethers::contract::parse_log(log)?;
-            self.event_processor.process_event(event).await?
+        loop {
+            self.recreate_client(
+                &mut client,
+                &mut recreate_client,
+                MAX_RETRIES,
+                WAIT_DURATION,
+            )
+            .await?;
```

Comment:
It's important to not that this function will return an error if we exhaust `MAX_RETRIES`, and by extension we could panic the downloader. By current settings, if we are not able to reconnect within 7 days, we'll panic.